src/optimizer/transforms/combine-repeating-patterns-transform.js

```diff
@@ -0,0 +1,206 @@
+      right = child;
+      left = alternative.getChild(startIndex);
+    } else {
+      right = NodePath.getForNode({
+        type: 'Alternative',
+        expressions: [...node.expressions.slice(index - i, index), child.node]
+      });
+
+      left = NodePath.getForNode({
+        type: 'Alternative',
+        expressions: [...node.expressions.slice(startIndex, index - i)]
+      });
+    }
+
+    if (right.hasEqualSource(left)) {
+      for (let j = 0; j < 2 * i + 1; j++) {
+        alternative.getChild(startIndex).remove();
+      }
+
+      child.replace({
+        type: 'Repetition',
+        expression: i === 0 ? right.node : {
+          type: 'Group',
+          capturing: false,
+          expression: right.node
+        },
+        quantifier: {
+          type: 'Quantifier',
+          kind: 'Range',
+          from: 2,
+          to: 2,
+          greedy: true
+        }
+      });
+      return startIndex;
+    }
+
+    i++;
+  }
+
+  return index;
+}
+
+// (?:abc){2}abc -> (?:abc){3}
+function combineWithPreviousRepetition(alternative, child, index) {
+  const {node} = alternative;
+
+  let i = 0;
+  while (i < index) {
+    let previousChild = alternative.getChild(i);
+
+    if (previousChild.node.type === 'Repetition' && previousChild.node.quantifier.greedy) {
+      let left = previousChild.getChild();
+      let right;
+
+      if (left.node.type === 'Group' && !left.node.capturing) {
+        left = left.getChild();
+      }
+
+      if (i + 1 === index) {
+        right = child;
+        if (right.node.type === 'Group' && !right.node.capturing) {
+          right = right.getChild();
+        }
+      } else {
+        right = NodePath.getForNode({
+          type: 'Alternative',
+          expressions: [...node.expressions.slice(i + 1, index + 1)]
+        });
+      }
+
+      if (left.hasEqualSource(right)) {
+
+        for (let j = i; j < index; j++) {
+          alternative.getChild(i + 1).remove();
+        }
+
+        increaseQuantifierByOne(previousChild.node.quantifier);
+
+        return i;
+      }
+    }
+
+    i++;
+  }
+  return index;
+}
+
+// abc(?:abc){2} -> (?:abc){3}
+function combineRepetitionWithPrevious(alternative, child, index) {
+  const {node} = alternative;
+
+  if (child.node.type === 'Repetition' && child.node.quantifier.greedy) {
+    let right = child.getChild();
+    let left;
+
+    if (right.node.type === 'Group' && !right.node.capturing) {
+      right = right.getChild();
+    }
+
+    let rightLength;
+    if (right.node.type === 'Alternative') {
+      rightLength = right.node.expressions.length;
+      left = NodePath.getForNode({
+        type: 'Alternative',
+        expressions: [...node.expressions.slice(index - rightLength, index)]
+      });
+    } else {
+      rightLength = 1;
+      left = alternative.getChild(index - 1);
+      if (left.node.type === 'Group' && !left.node.capturing) {
+        left = left.getChild();
+      }
+    }
+
+    if (left.hasEqualSource(right)) {
```

Comment:
BTW, later we should probably start caching the sources of the NodePaths in some registry for performance improvements (and checking the cached source in `hasEqualSource` implementation).

Reply:
@DmitrySoshnikov Aren't you afraid it will be hard to keep the source in sync with the node? I mean, changing any property of the node will change the source.

Reply:
@Golmote, yeah, good point, this can quickly go unsync. For now that's said it should be fine in general, and we can just cache the source into variables in case in transforms themselves (and recalculate it if nodes change).